File: backend/src/main/java/com/crm/security/RefreshTokenService.java
package com.crm.security;

import java.time.Instant;
import java.util.UUID;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.crm.entity.Benutzer;
import com.crm.entity.RefreshToken;
import com.crm.exception.TokenRefreshException;
import com.crm.repository.RefreshTokenRepository;

@Service
@Transactional
public class RefreshTokenService {

    private final RefreshTokenRepository refreshTokenRepository;
    private final long refreshTokenExpiration;

    public RefreshTokenService(
            RefreshTokenRepository refreshTokenRepository,
            @Value("${jwt.refresh-token-expiration}") long refreshTokenExpiration) {
        this.refreshTokenRepository = refreshTokenRepository;
        this.refreshTokenExpiration = refreshTokenExpiration;
    }

    public RefreshToken createRefreshToken(Benutzer benutzer) {
        RefreshToken refreshToken = new RefreshToken();
        refreshToken.setBenutzer(benutzer);
        refreshToken.setToken(UUID.randomUUID().toString());
        refreshToken.setExpiryDate(Instant.now().plusMillis(refreshTokenExpiration));
        refreshToken.setCreatedAt(Instant.now());
        return refreshTokenRepository.save(refreshToken);
    }

    @Transactional(readOnly = true)
    public RefreshToken validateRefreshToken(String token) {
        RefreshToken refreshToken = refreshTokenRepository.findByToken(token)
                .orElseThrow(() -> new TokenRefreshException("Ungültiger Refresh-Token"));

        if (refreshToken.getExpiryDate().isBefore(Instant.now())) {
            refreshTokenRepository.delete(refreshToken);
            throw new TokenRefreshException("Refresh-Token abgelaufen");
        }

        return refreshToken;
    }

    public void deleteByBenutzer(Benutzer benutzer) {
        refreshTokenRepository.deleteByBenutzer(benutzer);
    }
}

File: ciam/src/main/kotlin/com/crm/ciam/security/RefreshTokenService.kt
package com.crm.ciam.security

import com.crm.ciam.entity.Benutzer
import com.crm.ciam.entity.RefreshToken
import com.crm.ciam.exception.TokenRefreshException
import com.crm.ciam.repository.RefreshTokenRepository
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant
import java.util.UUID

@Service
@Transactional
class RefreshTokenService(
    private val refreshTokenRepository: RefreshTokenRepository,
    @Value("\${jwt.refresh-token-expiration}") private val refreshTokenExpiration: Long
) {

    fun createRefreshToken(benutzer: Benutzer): RefreshToken {
        val refreshToken = RefreshToken()
        refreshToken.benutzer = benutzer
        refreshToken.token = UUID.randomUUID().toString()
        refreshToken.expiryDate = Instant.now().plusMillis(refreshTokenExpiration)
        refreshToken.createdAt = Instant.now()
        return refreshTokenRepository.save(refreshToken)
    }

    @Transactional(readOnly = true)
    fun validateRefreshToken(token: String): RefreshToken {
        val refreshToken = refreshTokenRepository.findByToken(token)
            .orElseThrow { TokenRefreshException("Ungültiger Refresh-Token") }

        if (refreshToken.expiryDate!!.isBefore(Instant.now())) {
            refreshTokenRepository.delete(refreshToken)
            throw TokenRefreshException("Refresh-Token abgelaufen")
        }

        return refreshToken
    }

    fun deleteByBenutzer(benutzer: Benutzer) {
        refreshTokenRepository.deleteByBenutzer(benutzer)
    }
}
